generator/internal/parser/httprule/http_rule_parser.go

```diff
@@ -0,0 +1,382 @@
+	plainSegments := strings.Split(segmentsString, string(slash))
+	parsedSegments := make([]*Segment, 0, len(plainSegments))
+	for _, plainSegment := range plainSegments {
+		parsedSegment, err := parsePlainSegment(plainSegment)
+		if err != nil {
+			return nil, err
+		}
+		parsedSegments = append(parsedSegments, parsedSegment)
+	}
+	return parsedSegments, nil
+}
+
+func parseFieldPath(fieldPathString string) ([]*Identifier, error) {
+	identifiers := strings.Split(fieldPathString, string(identSep))
+	parsedIdentifiers := make([]*Identifier, 0, len(identifiers))
+	for _, identifier := range identifiers {
+		identifier, err := parseIdentifier(identifier)
+		if err != nil {
+			return nil, err
+		}
+		parsedIdentifiers = append(parsedIdentifiers, identifier)
+	}
+	return parsedIdentifiers, nil
+}
+
+func parsePlainSegment(plainSegment string) (*Segment, error) {
+	if plainSegment == string(star) {
+		return &Segment{Match: &Match{}}, nil
+	} else if plainSegment == string(star)+string(star) {
+		return &Segment{MatchRecursive: &MatchRecursive{}}, nil
+	} else {
+		literal, err := parseLiteral(plainSegment)
+		if err != nil {
+			return nil, err
+		}
+		return &Segment{Literal: literal}, nil
+	}
+}
+
+const (
+	hexStart   = '%'
+	hexdig     = "0123456789ABCDEFabcdef"
+	digit      = "0123456789"
+	alpha      = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
+	subDelims  = "!$&'()*+,;="
+	unreserved = alpha + digit + "-._~"
+	pchar      = unreserved + subDelims + ":@"
+)
+
+// parseLiteral validates that the provided string conforms to the LITERAL definition, and returns a Literal type if it does.
+func parseLiteral(literal string) (*Literal, error) {
+	if len(literal) < 1 {
+		return nil, fmt.Errorf("invalid literal, expected at least one character: %s", literal)
+	}
+	i := 0
+	for i < len(literal) {
+		if strings.ContainsRune(pchar, rune(literal[i])) {
+			i++
+		} else if literal[i] == hexStart {
+			if i+2 >= len(literal) {
+				return nil, fmt.Errorf("invalid literal, expected at least 2 characters after the '%%': %s", literal)
+			}
+			if !strings.ContainsRune(hexdig, rune(literal[i+1])) || !strings.ContainsRune(hexdig, rune(literal[i+2])) {
+				return nil, fmt.Errorf("invalid literal: %s", literal)
+			}
+			i += 3
+		} else {
+			return nil, fmt.Errorf("invalid literal: %s", literal)
+		}
+	}
+	return (*Literal)(&literal), nil
+}
+
+func parseIdentifier(identifier string) (*Identifier, error) {
+	if len(identifier) < 1 {
+		return nil, fmt.Errorf("invalid identifier, expected at least one character: %s", identifier)
+	}
+	if !strings.ContainsRune(alpha, rune(identifier[0])) {
+		return nil, fmt.Errorf("invalid identifier, expected it to start with a letter: %s", identifier)
+	}
+	i := 1
+	for i < len(identifier) {
+		if strings.ContainsRune(alpha+digit+"_", rune(identifier[i])) {
+			i++
+		} else {
+			return nil, fmt.Errorf("invalid identifier, rune '%q' is not valid in an identifier: %s", identifier[i], identifier)
+		}
+	}
```

Comment:
I think you can write this as:

```suggestion
        if i := strings.IndexAny(identifier, alpha+digit+"_"; i != -1 {
            return nil, fmt.Errorf("invalid identifier, rune '%q' is not valid in an identifier: %s", identifier[i], identifier)
	}
```

Reply:
On second thought, maybe not, but `strings.IndexFunc(identifier, func(r rune) bool { return ! strings.ContainsRune(alpha + digit + "_", r); })` may be closer to what you want.

Reply:
I left the previous comment on a stale screen and hadn't seen your second suggestion. That one did work, and looks neat!

Reply:
updated